server/converters/diffexp_precompute.py

```diff
@@ -0,0 +1,144 @@
+    if os.path.exists(diffexp_group):
+        print(f"{diffexp_group} exists")
+        if args.overwrite:
+            print(f"{diffexp_group} exists, will overwrite")
+            shutil.rmtree(diffexp_group)
+        else:
+            print(f"{diffexp_group} exists, use '--overwrite' to overwrite existing files")
+            sys.exit(1)
+
+    shape = adaptor.get_shape()
+    attrs = adaptor.get_obs_columns()
+
+    # determine which obs values to precompute, and how many.
+    num_precompute = 0
+    attr_precompute = []
+    for attr in attrs:
+        values = adaptor.query_obs_array(attr)
+        uvalues = np.unique(values)
+        ngroups = len(uvalues)
+        if ngroups > args.max_groups:
+            print(f"Skipping '{attr}': too many groups {ngroups} > {args.max_groups}")
+            continue
+        if ngroups < args.min_groups:
+            print(f"Skipping '{attr}': too few groups {ngroups} < {args.min_groups}")
+            continue
+
+        print(f"Queueing '{attr}': num groups {ngroups}")
+        num_precompute += ngroups
+        attr_precompute.append(attr)
+
+    hashkey = np.empty((num_precompute,), dtype=object)
+    metakey = np.empty((num_precompute, 2), dtype=object)
+    mean_var = np.zeros((2, num_precompute, shape[1]), dtype=np.float64)
+
+    print()
+    num_precompute = precompute(adaptor, attr_precompute, hashkey, metakey, mean_var, args.max_rows)
+
+    np.resize(hashkey, (num_precompute,))
+    np.resize(metakey, (num_precompute, 2))
+    np.resize(mean_var, (2, num_precompute, shape[1]))
+
+    # check if group exists
+    tiledb.group_create(diffexp_group, CxgAdaptor.tiledb_ctx)
+    hash_version = compute_indices_hash_version()
+    save_array(diffexp_group, "hashkey", hashkey, dict(hash_version=hash_version))
+    save_array(diffexp_group, "metakey", metakey)
+    save_array(diffexp_group, "mean_var", mean_var)
+
+    # TODO save metadata about the hash function version, in case it gets changed
+
+
+def precompute(adaptor, attrnames, hashkey, metakey, mean_var, max_rows):
+    hash_version = compute_indices_hash_version()
+    pindex = 0
+    for attr in attrnames:
+        values = adaptor.query_obs_array(attr)
+        uvalues = np.unique(values)
+        ngroups = len(uvalues)
+        print(f"Processing '{attr}': num groups = {ngroups}")
+        for i in range(ngroups):
+            val = uvalues[i]
+            groupindices = np.where(values == val)[0]
+            grouplen = len(groupindices)
+            if grouplen > max_rows:
+                print(f" '{attr}'='{str(val)}': num indices = {grouplen} > {max_rows}: skipping")
+                continue
+            else:
+                print(f" '{attr}'='{str(val)}': num indices = {grouplen}")
+            try:
+                tmean, tvariance = compute_mean_var(adaptor, groupindices)
+            except ComputeError as e:
+                print(f"   error: {str(e)}")
+                continue
+            mean_var[0][pindex] = tmean
```

Comment:
what is the benefit of strong mean and var in a single (2, N) array, rather than creating a two attribute array?  The latter has the benefit of more easily being expanded when you decide you want to compute additional summary metrics for these row groups - you just add another tiledb attribute to the array.

it also makes the code a lot easier to understand.